File: projects/repo-intake-fbabaei-microsoft-azure-architecture-factory-20260425012922/projects/mdr-support-20260421002041/src/mdr_support/services/session_service.py
"""Session service: records chat turns per document."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from threading import Lock


@dataclass
class ChatTurn:
    at: str
    field: str
    answer: str


@dataclass
class Session:
    document_id: str
    turns: list[ChatTurn] = field(default_factory=list)


class SessionService:
    def __init__(self) -> None:
        self._lock = Lock()
        self._sessions: dict[str, Session] = {}

    def record_turn(self, document_id: str, field: str, answer: str) -> None:
        with self._lock:
            session = self._sessions.setdefault(document_id, Session(document_id=document_id))
            session.turns.append(ChatTurn(at=datetime.now(timezone.utc).isoformat(), field=field, answer=answer))

    def get(self, document_id: str) -> Session | None:
        with self._lock:
            return self._sessions.get(document_id)

File: projects/repo-intake-fbabaei-microsoft-azure-architecture-factory-20260425012922/projects/mdr-support-20260421003634/src/Services/SessionService.cs
namespace GeneratedApi.Services;

using System.Collections.Concurrent;

public record ChatTurn(DateTimeOffset At, string Field, string Answer);

public record Session(string DocumentId, List<ChatTurn> Turns);

public class SessionService
{
    private readonly ConcurrentDictionary<string, Session> _sessions = new();

    public void RecordTurn(string documentId, string field, string answer)
    {
        var session = _sessions.GetOrAdd(documentId, id => new Session(id, new List<ChatTurn>()));
        lock (session.Turns)
        {
            session.Turns.Add(new ChatTurn(DateTimeOffset.UtcNow, field, answer));
        }
    }

    public Session? Get(string documentId)
        => _sessions.TryGetValue(documentId, out var session) ? session : null;
}
